doc: Enabling ACRN on top of QEMU/KVM (#4802)
projectacrn/acrn-hypervisor: doc/tutorials/acrn_on_qemu.rst

.. _acrn_on_qemu:

Enable ACRN over QEMU/KVM
#########################

Goal of this document is to bring-up ACRN as a nested Hypervisor on top of QEMU/KVM
with basic functionality such as running Service VM (SOS) and User VM (UOS) for primarily 2 reasons,

1. In order for the users to evaluate ACRN.
2. Make ACRN platform agnostic and remove per hardware platform configurations setup overhead.

This setup was tested with the following configuration,

> 1. ACRN Hypervisor: Release_1.6 Branch (Commit ID: d661d444c62b7d9785c49695934f543ef3b7b679

Author: This is in release_1.6 branch. But here is a question. Is the commit id even needed? I have tested on the master branch and it works as well. We can remove this line itself and tell user to use the latest.

> 1. ACRN Hypervisor: Release_1.6 Branch (Commit ID: d661d444c62b7d9785c49695934f543ef3b7b679

Reviewer: My preference is always to give at least one known good version (i.e. one that we have tested thoroughly). People are welcome to try the latest master but if things go wrong, I want them to be able to at least confirm whether it's something that went wrong on their side vs. potentially a regression/bug in our code.

> 1. ACRN Hypervisor: Release_1.6 Branch (Commit ID: d661d444c62b7d9785c49695934f543ef3b7b679

Author: Okay added tag, and commit ID.

2. ACRN Kernel: 5.4.28 (Commit ID: a8cd22f49f0b2b56e526150fe0aaa9118edfcede)
3. QEMU emulator version 4.2.0
4. SOS/UOS is ubuntu18.04

Pre-Requisites
**************
1. Make sure the platform supports Intel VMX as well as VT-d technologies. On ubuntu18.04.4, this

> can be checked by installing ``cpu-checker`` tool. If the output displays **KVM acceleration can be used**

Reviewer 2: subsequent lines of a numbered list must be indented (3 spaces)

Author: done.

the platform supports it.

  .. code-block:: none

     acrn@host:~$ kvm-ok
     INFO: /dev/kvm exists
     KVM acceleration can be used

2. Make sure KVM and the following utilities are installed.

  .. code-block:: none

     sudo apt update && sudo apt upgrade -y
     sudo apt install qemu-kvm libvirt-bin virtinst -y

Prepare SOS (L1 Guest)
**********************

> 1. Use virt-install command to create SOS.

Reviewer: ```suggestion
1. Use ``virt-install`` command to create SOS.
```

Author: done.

   .. code-block:: none

      virt-install \
      --connect qemu:///system \
      --name ACRNSOS \
      --machine q35 \
      --cpu host-passthrough,+invtsc \
      --ram 4096 \
      --disk path=/var/lib/libvirt/images/acrnsos.img,size=32 \
      --vcpus 4 \
      --virt-type kvm \
      --os-type linux \
      --os-variant ubuntu18.04 \
      --graphics none \
      --clock offset=utc,tsc_present=yes,kvmclock_present=no \
      --qemu-commandline="-machine kernel-irqchip=split -device intel-iommu,intremap=on,caching-mode=on,aw-bits=48" \
      --location 'http://archive.ubuntu.com/ubuntu/dists/bionic/main/installer-amd64/' \
      --extra-args "console=tty0 console=ttyS0,115200n8"

2. Walk through the installation steps as prompted. Few things to note,

>    a. Make sure to install an open-ssh server so that once the installation is complete, we can ssh into the system.

Reviewer: ```suggestion
   a. Make sure to install an OpenSSH server so that once the installation is complete, we can SSH into the system.
```

Author: done.

      .. figure:: images/acrn_qemu_1.png
         :align: center

   b. We use GRUB to boot ACRN, so please install it when prompted.

      .. figure:: images/acrn_qemu_2.png
         :align: center

3. To login to the SOS guest, find the IP address of the guest to SSH. This can be done via the

> virsh command as shown below,

Reviewer 2: indent (3 spaces)

Author: done.

   .. figure:: images/acrn_qemu_3.png
      :align: center

4. Once ACRN hypervisor is enabled, the above virsh command might not display the IP. So enable Serial console by,

   .. code-block:: none

      sudo systemctl enable serial-getty@ttyS0.service
      sudo systemctl start serial-getty@ttyS0.service

> 5. Enable grub menu to choose between Ubuntu Vs ACRN hypervisor. Modify :file:`/etc/default/grub` and edit below entries,

Reviewer: ```suggestion
5. Enable GRUB menu to choose between Ubuntu vs. ACRN hypervisor. Modify :file:`/etc/default/grub` and edit below entries,
```

Author: done.

      .. code-block:: none

         GRUB_TIMEOUT_STYLE=menu
         GRUB_TIMEOUT=5
         GRUB_CMDLINE_LINUX_DEFAULT=""
         GRUB_GFXMODE=text

> 6. sudo update-grub

Reviewer: ```suggestion
6. ``sudo update-grub``
```

Author: modified as "6. Update GRUB changes by ``sudo update-grub``"

7. Once the above steps are done, SOS guest can also be launched using, ``virsh start ACRNSOS --console``. Make sure to use the domain name

> you used while creating the VM instead of ``ACRNSOS``.

Reviewer 2: indent (3 spaces)

Author: done.

This concludes setting up of SOS and preparing it to boot ACRN hypervisor.

Install ACRN Hypervisor
***********************

1. Clone ACRN hypervisor.

   .. code-block:: none

      $ git clone https://github.com/projectacrn/acrn-hypervisor.git

>       $ cd acrn-hypervisor/

Reviewer: Add a note here to optionnally checkout the known good version. My suggestion would be to phrase such as to let the use use the latest master if he wants, or he can check out the version that was tested at the time we wrote this doc.

Author: Yeah, even I think we can just let the user pick the latest stable release or master and not suggest any version.

Reviewer: Let's provide a known good version, also invite to try master if the user wishes so

Author: done.

2. Use the following command to build ACRN for QEMU,

   .. code-block:: none

>       make all BOARD_FILE=../acrn-hypervisor/misc/acrn-config/xmls/board-xmls/qemu.xml  SCENARIO_FILE=../acrn-hypervisor/misc/acrn-config/xmls/config-xmls/qemu/sdc.xml

Reviewer: Why the `../acrn-hypervisor` here? This takes you back to where you are, doesn't it?

Author: good catch. It is a typo, it should be from misc.

 For more details, refer to :ref:`getting-started-building`.

> 3. Copy acrn.32.out from build/hypervisor to SOS guest /boot/ directory.

Reviewer: ```suggestion
3. Copy ``acrn.32.out`` from ``build/hypervisor`` to SOS guest ``/boot/`` directory.
```

Author: done.

4. Update the SOS kernel so as to include the virtio-blk driver. UOS (L2 guest) uses virtio-blk

> driver to mount rootfs.

Reviewer 2: indent (3 spaces)

Author: done.

>    a. Clone the repo, https://github.com/projectacrn/acrn-kernel.git.

Reviewer: ```suggestion
   a. Clone the repo: ``git clone https://github.com/projectacrn/acrn-kernel``
```

Author: done.

>    b. cd acrn-kernel

Reviewer: Insert note below to optionally checkout the known good version.

Author: Reworded and created a code block similar to earlier writeup where I say how to clone acrn hypervisor.

   c. Copy kernel_config_uefi_sos to .config
   d. make olddefconfig
   e. make menuconfig

      .. figure:: images/acrn_qemu_4.png
         :align: center

>    f. make

Reviewer: No need to build and install the kernel modules?

Author: Since all the drivers are part of the kernel, I didn't need to install the modules.

Reviewer: Can we add a note explaining that?

>    g. Copy bzImage go the SOS guest /boot/ directory

Reviewer: ```suggestion
   g. Copy ``arch/x86/boot/bzImage`` to the SOS guest ``/boot/`` directory
```

Author: done.

5. Update Ubuntu GRUB to boot ACRN hypervisor and load ACRN Kernel Image. Append the following

> configuration to the :file:`/etc/grub.d/40_custom`,

Reviewer 2: indent (3 spaces)

Author: done.

   .. code-block:: none

      menuentry 'ACRN hypervisor' --class ubuntu --class gnu-linux --class gnu --class os $menuentry_id_option 'gnulinux-simple-e23c76ae-b06d-4a6e-ad42-46b8eedfd7d3' {
         recordfail
         load_video
         gfxmode $linux_gfx_mode
         insmod gzio
         insmod part_msdos
         insmod ext2

         echo 'Loading ACRN hypervisor with SDC scenario ...'
         multiboot --quirk-modules-after-kernel /boot/acrn.32.out

>          module /boot/bzImage_sos Linux_bzImage

Reviewer: Is there an expectation here that the bzImage` you took above be copied under one of those names? (I.e. `bzImage_sos` or `Linux_bzImage`)

Author: We are creating 2 bzImage one for SOS and UOS in the user guide. Here I mean bzImage that is used for SOS, so named it bzImage_sos. Linux_bzImage is the kernel mod tag and should be used as stated in the doc. This should not be changed.

      }

> 6. Update grub ``sudo update-grub``.

Reviewer: ```suggestion
6. Update GRUB: ``sudo update-grub``.
```

Author: done.

> 7. Shutdown the guest and relaunch using, virsh start ACRNSOS --console and select ACRN hypervisor from grub menu to launch SOS running on top of ACRN.

Reviewer: ```suggestion
7. Shutdown the guest and relaunch using, ``virsh start ACRNSOS --console`` and select ACRN hypervisor from GRUB menu to launch SOS running on top of ACRN.
```

Author: done.

> This can be verified from dmesg as shown below,

Reviewer: ```suggestion
This can be verified from ``dmesg`` as shown below,
```

Author: done.

> This can be verified from dmesg as shown below,

Reviewer 2: indent (3 spaces)

Author: done.

   .. code-block:: none

      guestl1@ACRNSOS:~$ dmesg | grep ACRN
      [    0.000000] Hypervisor detected: ACRN
      [    2.337176] ACRNTrace: Initialized acrn trace module with 4 cpu
      [    2.368358] ACRN HVLog: Initialized hvlog module with 4 cpu
      [    2.727905] systemd[1]: Set hostname to <ACRNSOS>.

8. When shutting down, make sure to cleanly destroy the SOS VM to prevent crashes in subsequent boots. This can be done using,

      .. code-block:: none

>          virsh destory ACRNSOS, where ACRNSOS is the virsh domain name.

Reviewer: ```suggestion
         virsh destory ACRNSOS # where ACRNSOS is the virsh domain name.
```

Author: done.

SOS Networking updates for UOS
******************************
If you are planning to enable networking for UOS(L2 guest), the following steps might be helpful.

1. Edit your :file:`/etc/netplan/01-netcfg.yaml` file to add acrn-br0 as below,

   .. code-block:: none

      network:
         version: 2
         renderer: networkd
         ethernets:
            enp1s0:
               dhcp4: no
         bridges:
            acrn-br0:
               interfaces: [enp1s0]
               dhcp4: true
               dhcp6: no

2. Apply the new network configuration by,

>       a. cd /etc/netplan

Reviewer: Probably better put in a `code-block` section

Author: done.

      #. sudo netplan generate
      #. sudo netplan apply

3. Create a tap interface (tap0) and add the tap interface as part of the acrn-br0.

> Copy the following files to :file:`/usr/lib/system/network`,

Reviewer 2: indent (lines 205, 207-8) three spaces

Author: done.

- `50-acrn.network <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/misc/acrnbridge/acrn.network>`__
- `50-tap0.netdev <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/misc/acrnbridge/tap0.netdev>`__

4. Restart ACRNSOS guest (L1 guest) to complete the setup and start with bring-up of UOS

Bring-up UOS (L2 Guest)
***********************
1. Build the device-model, using make devicemodel and copy acrn-dm to ACRNSOS guest (L1 guest) directory ``/usr/bin/acrn-dm``

   .. note::
      It should be already built as part of .._Install ACRN Hypervisor.

2. On the ACRNSOS guest, install shared libraries for acrn-dm (if not already installed).

   .. code-block:: none

      sudo apt-get install libpciaccess-dev

3. Install latest `IASL tool <https://acpica.org/downloads>`_ and copy the binary to ``/usr/sbin/iasl``.

> For this setup, used IASL 20200326 version but anything after 20190215 should be good.

Reviewer 2: indent (3 spaces)

Author: done.

4. Build ACRN UOS Kernel by,

>       a. Clone the repo, https://github.com/projectacrn/acrn-kernel.git

Reviewer: ```suggestion
      a. Clone the repo: ``git clone https://github.com/projectacrn/acrn-kernel``
```

Reviewer: Good candidate to be put in a `code-block` section

Author: done.

      b. cd acrn-kernel

>       c. Copy kernel_config_uos to .config

Reviewer: Missing a `make olddefconfig` here I believe

Author: It works even without doing make olddefconfig. Do you want to add this for consistency?

>       d. make

Reviewer: No need to build and install the kernel modules?

Author: Since the needed drivers are part of the kernel, I didn't need to install the modules

Reviewer: Same here, can we add a note explaining that?

>       e. Copy bzImage go the ACRNSOS (L1 guest). Need this to launch the UOS (L2 guest)

Reviewer: ```suggestion
      e. Copy ``arch/x86/boot/bzImage`` to the ACRNSOS (L1 guest). Need this to launch the UOS (L2 guest)
```

And please add the destination name is a specific one is expected

Author: done.

5. Build ubuntu.img using :ref:`build-the-ubuntu-kvm-image` and copy it to the ACRNSOS (L1 Guest).

> Alternatively you can also use virt-install to create a UOS image similar to ACRNSOS as shown below,

Reviewer 2: indent (3 spaces)

Author: done.

   .. code-block:: none

      virt-install \
      --name UOS \
      --ram 2048 \
      --disk path=/var/lib/libvirt/images/UOSUbuntu.img,size=8 \
      --vcpus 2 \
      --virt-type kvm \
      --os-type linux \
      --os-variant ubuntu18.04 \
      --graphics none \
      --location 'http://archive.ubuntu.com/ubuntu/dists/bionic/main/installer-amd64/' \
      --extra-args "console=tty0 console=ttyS0,115200n8"

   .. note::

>       Image at /var/lib/libvirt/images/UOSUbuntu.img is qcom2 image. Convert it to raw image using, ``qemu-img convert -f qcow2 UOSUbuntu.img -O raw UOS.img``

Reviewer: ```suggestion
      Image at ``/var/lib/libvirt/images/UOSUbuntu.img`` is a qcow2 image. Convert it to raw image using, ``qemu-img convert -f qcow2 UOSUbuntu.img -O raw UOS.img``
```

Author: done.

Author: done.

> 6. Launch UOS using `launch_ubuntu.sh <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/doc/tutorial/launch_ubuntu.sh>`_ script. Make sure to update with your ubuntu image and rootfs

Reviewer: If that script is in the code repo we've cloned above, it's better to use it rather than pointing a master (which will change over time)

Author: No this is not part of sample scripts we have in master. I can add it there if that is okay but we must ensure it is maintained.

Reviewer: This would be my preference. Regardless of where it is our repo, it needs to be maintained. So from that point of view, I'd rather have it where all the others already are.

Author: Currently, we are supporting script files for 3 platforms in the devicemodel/samples directory. launch_ubuntu.sh doesn't fit in any of those as it is not platform-specific. So just placing this in the devicemodel/samples directory.

.. _acrn_on_qemu:

Enable ACRN over QEMU/KVM
#########################

Goal of this document is to bring-up ACRN as a nested Hypervisor on top of QEMU/KVM
with basic functionality such as running Service VM (SOS) and User VM (UOS) for primarily 2 reasons,

1. In order for the users to evaluate ACRN.
2. Make ACRN platform agnostic and remove per hardware platform configurations setup overhead.

This setup was tested with the following configuration,

> 1. ACRN Hypervisor: tag acrn-2020w19.5-140000p (Commit ID: b0ae9cfa2b374d3faad9f890ff61e88f97205ac8)

Reviewer 2: use a bullet list if there's no reason to have numbered items:
```
- ACRN Hypervisor ....
- ACRN Kernel: ....

```

   the platform supports it.

  .. code-block:: none

     acrn@host:~$ kvm-ok
     INFO: /dev/kvm exists
     KVM acceleration can be used

2. Make sure KVM and the following utilities are installed.

  .. code-block:: none

     sudo apt update && sudo apt upgrade -y
     sudo apt install qemu-kvm libvirt-bin virtinst -y

Prepare Service VM (L1 Guest)
*****************************
1. Use ``virt-install`` command to create Service VM.

   .. code-block:: none

      virt-install \
      --connect qemu:///system \
      --name ACRNSOS \
      --machine q35 \
      --cpu host-passthrough,+invtsc \
      --ram 4096 \
      --disk path=/var/lib/libvirt/images/acrnsos.img,size=32 \
      --vcpus 4 \
      --virt-type kvm \
      --os-type linux \
      --os-variant ubuntu18.04 \
      --graphics none \
      --clock offset=utc,tsc_present=yes,kvmclock_present=no \
      --qemu-commandline="-machine kernel-irqchip=split -device intel-iommu,intremap=on,caching-mode=on,aw-bits=48" \
      --location 'http://archive.ubuntu.com/ubuntu/dists/bionic/main/installer-amd64/' \
      --extra-args "console=tty0 console=ttyS0,115200n8"

> 2. Walk through the installation steps as prompted. Few things to note,

Reviewer 2: change to ``Here are a few things to note:``

   a. Make sure to install an OpenSSH server so that once the installation is complete, we can SSH into the system.

      .. figure:: images/acrn_qemu_1.png
         :align: center

   b. We use GRUB to boot ACRN, so please install it when prompted.

      .. figure:: images/acrn_qemu_2.png
         :align: center

3. To login to the Service VM guest, find the IP address of the guest to SSH. This can be done via the
   virsh command as shown below,

>    .. figure:: images/acrn_qemu_3.png

Reviewer 2: You can also do this using a ``.. code-block:: console`` to have a black background with white text, instead of providing a picture (that we can't edit).

Author: In addition to command, I want to show the output to the user. Here code block might not help.

      :align: center

> 4. Once ACRN hypervisor is enabled, the above virsh command might not display the IP. So enable Serial console by,

Reviewer 2: This tip should probably be before the previous one (showing the virsh command).  If you leave it here, you should also say to run the virsh command again after enabling the serial console.

Author: This command need not be run after enabling the serial console. The point I am trying to make here is that virsh commands like this will not work once ACRN hypervisor is enabled. So we need to have console to display grub options.

   .. code-block:: none

      sudo systemctl enable serial-getty@ttyS0.service
      sudo systemctl start serial-getty@ttyS0.service

5. Enable GRUB menu to choose between Ubuntu vs ACRN hypervisor. Modify :file:`/etc/default/grub` and edit below entries,

      .. code-block:: none

         GRUB_TIMEOUT_STYLE=menu
         GRUB_TIMEOUT=5
         GRUB_CMDLINE_LINUX_DEFAULT=""
         GRUB_GFXMODE=text

> 6. Update GRUB changes by ``sudo update-grub``

Reviewer 2: better to put commands into a code-block (and with a leading ``$``)

Author: The earlier are modifications to a file and not a command. So cannot club this together in a code block. Also, having a code block just for this command might not look good.

7. Once the above steps are done, Service VM guest can also be launched using, ``virsh start ACRNSOS --console``. Make sure to use the domain name
   you used while creating the VM instead of ``ACRNSOS``.

This concludes setting up of Service VM and preparing it to boot ACRN hypervisor.

.. _install_acrn_hypervisor:

Install ACRN Hypervisor
***********************

1. Please clone ACRN repo with ``tag: acrn-2020w19.5-140000p`` or latest master. Below steps show our tested version,

   .. code-block:: none

      $ git clone https://github.com/projectacrn/acrn-hypervisor.git
      $ cd acrn-hypervisor/
      $ git fetch --all --tags --prune
      $ git checkout tags/acrn-2020w19.5-140000p -b acrn_on_qemu

2. Use the following command to build ACRN for QEMU,

   .. code-block:: none

      make all BOARD_FILE=./misc/acrn-config/xmls/board-xmls/qemu.xml  SCENARIO_FILE=./misc/acrn-config/xmls/config-xmls/qemu/sdc.xml

 For more details, refer to :ref:`getting-started-building`.

> 3. Copy ``acrn.32.out`` from ``build/hypervisor`` to Service VM guest ``/boot/`` directory.

Reviewer 2: Use a code-block with commands instead of a sentence.

Author: The user can build in a different system and copy it SUT. So cannot explicitly put this as command and so decided to give it as an instruction.

> 4. Build the Service VM kernel so as to include the virtio-blk driver. User VM (L2 guest) uses virtio-blk

Reviewer 2: change ``to as to include`` to ``that includes``

   driver to mount rootfs.

   a. Please clone the latest stable version or master as show below,

      .. code-block:: none

         $ git clone https://github.com/projectacrn/acrn-kernel
         $ cd acrn-kernel

   b. Copy kernel_config_uefi_sos to .config
   c. make olddefconfig
   d. make menuconfig

      .. figure:: images/acrn_qemu_4.png
         :align: center

   f. make
   g. Copy ``arch/x86/boot/bzImage`` to the Service VM /boot/ directory and rename it to ``bzImage_sos``.

   .. note::

>       Since the needed drivers are part of the Service VM kernel itself, we can avoid installing the modules.

Reviewer 2: Better would be:
```
The Service VM kernel contains all needed drivers so you won't need to install extra kernel modules.
```

         recordfail
         load_video
         gfxmode $linux_gfx_mode
         insmod gzio
         insmod part_msdos
         insmod ext2

         echo 'Loading ACRN hypervisor with SDC scenario ...'
         multiboot --quirk-modules-after-kernel /boot/acrn.32.out
         module /boot/bzImage_sos Linux_bzImage
      }

6. Update GRUB ``sudo update-grub``.

7. Shutdown the guest and relaunch using, ``virsh start ACRNSOS --console`` and select ACRN hypervisor from GRUB menu to launch Service VM running on top of ACRN.
   This can be verified from ``dmesg`` as shown below,

   .. code-block:: none

      guestl1@ACRNSOS:~$ dmesg | grep ACRN
      [    0.000000] Hypervisor detected: ACRN
      [    2.337176] ACRNTrace: Initialized acrn trace module with 4 cpu
      [    2.368358] ACRN HVLog: Initialized hvlog module with 4 cpu
      [    2.727905] systemd[1]: Set hostname to <ACRNSOS>.

8. When shutting down, make sure to cleanly destroy the Service VM to prevent crashes in subsequent boots. This can be done using,

>       .. code-block:: none

Reviewer 2: excessive indenting

>          virsh destory ACRNSOS # where ACRNSOS is the virsh domain name.

Reviewer 2: probably should be ``destroy``

Service VM Networking updates for User VM
*****************************************

> If you are planning to enable networking for User VM(L2 guest), the following steps might be helpful.

Reviewer 2: Better would be:
```
Follow these steps to enable networking for the User VM (L2 guest):
```

1. Edit your :file:`/etc/netplan/01-netcfg.yaml` file to add acrn-br0 as below,

   .. code-block:: none

      network:
         version: 2
         renderer: networkd
         ethernets:
            enp1s0:
               dhcp4: no
         bridges:
            acrn-br0:
               interfaces: [enp1s0]
               dhcp4: true
               dhcp6: no

2. Apply the new network configuration by,

>       .. code-block:: none

Reviewer 2: ReST is sensitive to extra spaces, so just indent the code block directive by three spaces (not six).  There are some other cases in this doc with excess indentation too.

         $ cd /etc/netplan
         $ sudo netplan generate
         $ sudo netplan apply

3. Create a tap interface (tap0) and add the tap interface as part of the acrn-br0.

>    Copy files ``misc/acrnbridge/acrn.network`` and ``misc/acrnbridge/tap0.netdev`` from the cloned ACRN repo to :file:`/usr/lib/system/network`.

Reviewer 2: Again, better to write a code-block with the commands than sentences here.

Author: The user can build in a different system and copy it SUT. So cannot explicitly put this as command and decided to give it as instructed.

   Rename them to ``50-acrn.network`` and ``50-tap0.netdev`` respectively.

> 4. Restart ACRNSOS guest (L1 guest) to complete the setup and start with bring-up of User VM

Reviewer 2: Can you provide the commands to do these steps?

Author: These are the steps. I have reworded. Let me know if this is more clear.

Bring-up User VM (L2 Guest)
***************************

> 1. Build the device-model, using make devicemodel and copy acrn-dm to ACRNSOS guest (L1 guest) directory ``/usr/bin/acrn-dm``

Reviewer 2: Give the commands (in a code-block) rather than describing the commands

Author: The user can build in a different system and copy it SUT. So cannot explicitly put this as command and decided to give it as instructed.

   .. note::
      It should be already built as part of :ref:`install_acrn_hypervisor`.

2. On the ACRNSOS guest, install shared libraries for acrn-dm (if not already installed).

   .. code-block:: none

>       sudo apt-get install libpciaccess-dev

Reviewer 2: leading ``$`` on commands

3. Install latest `IASL tool <https://acpica.org/downloads>`_ and copy the binary to ``/usr/sbin/iasl``.
   For this setup, used IASL 20200326 version but anything after 20190215 should be good.

4. Build ACRN User VM Kernel by,

>       a. Please clone the latest stable version or master as show below,

Reviewer 2: Excessive indenting

         .. code-block:: none

            $ git clone https://github.com/projectacrn/acrn-kernel
            $ cd acrn-kernel

>       b. Copy kernel_config_uos to .config

Reviewer 2: use a code-block with the commands

      c. make
      d. Copy ``arch/x86/boot/bzImage`` to  ACRNSOS (L1 guest) and rename this to ``bzImage_uos``. Need this to launch the User VM (L2 guest)

   .. note::
      Since the needed drivers are part of the User VM kernel itself, we can avoid installing the modules.

5. Build ubuntu.img using :ref:`build-the-ubuntu-kvm-image` and copy it to the ACRNSOS (L1 Guest).
   Alternatively you can also use virt-install to create a User VM image similar to ACRNSOS as shown below,

   .. code-block:: none

      virt-install \
      --name UOS \
      --ram 2048 \
      --disk path=/var/lib/libvirt/images/UOSUbuntu.img,size=8 \
      --vcpus 2 \
      --virt-type kvm \
      --os-type linux \
      --os-variant ubuntu18.04 \
      --graphics none \
      --location 'http://archive.ubuntu.com/ubuntu/dists/bionic/main/installer-amd64/' \
      --extra-args "console=tty0 console=ttyS0,115200n8"

   .. note::

>       Image at ``/var/lib/libvirt/images/UOSUbuntu.img`` is a qcow2 image. Convert it to raw image using, ``qemu-img convert -f qcow2 UOSUbuntu.img -O raw UOS.img``

Reviewer 2: use a code block for the commands...

Author: Do you mean to have a code block inside a note? I feel the note itself should be good.

      [    2.727905] systemd[1]: Set hostname to <ACRNSOS>.

8. When shutting down, make sure to cleanly destroy the Service VM to prevent crashes in subsequent boots. This can be done using,

      .. code-block:: none

         virsh destory ACRNSOS # where ACRNSOS is the virsh domain name.

Service VM Networking updates for User VM
*****************************************
If you are planning to enable networking for User VM(L2 guest), the following steps might be helpful.

1. Edit your :file:`/etc/netplan/01-netcfg.yaml` file to add acrn-br0 as below,

   .. code-block:: none

      network:
         version: 2
         renderer: networkd
         ethernets:
            enp1s0:
               dhcp4: no
         bridges:
            acrn-br0:
               interfaces: [enp1s0]
               dhcp4: true
               dhcp6: no

2. Apply the new network configuration by,

      .. code-block:: none

         $ cd /etc/netplan
         $ sudo netplan generate
         $ sudo netplan apply

3. Create a tap interface (tap0) and add the tap interface as part of the acrn-br0.
   Copy the following files to :file:`/usr/lib/system/network`,

>    - `50-acrn.network <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/misc/acrnbridge/acrn.network>`__

Reviewer: We can instruct the user to pick up these files from the local copy of the repo, the added benefit is that it will automatically be in sync with the version of ACRN that was checked out and built.

   - `50-tap0.netdev <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/misc/acrnbridge/tap0.netdev>`__

4. Restart ACRNSOS guest (L1 guest) to complete the setup and start with bring-up of User VM

Bring-up User VM (L2 Guest)
***************************
1. Build the device-model, using make devicemodel and copy acrn-dm to ACRNSOS guest (L1 guest) directory ``/usr/bin/acrn-dm``

   .. note::
      It should be already built as part of .._Install ACRN Hypervisor.

2. On the ACRNSOS guest, install shared libraries for acrn-dm (if not already installed).

   .. code-block:: none

      sudo apt-get install libpciaccess-dev

3. Install latest `IASL tool <https://acpica.org/downloads>`_ and copy the binary to ``/usr/sbin/iasl``.
   For this setup, used IASL 20200326 version but anything after 20190215 should be good.

4. Build ACRN User VM Kernel by,

      a. Please clone the latest stable version or master as show below,

         .. code-block:: none

            $ git clone https://github.com/projectacrn/acrn-kernel
            $ cd acrn-kernel

      b. Copy kernel_config_uos to .config
      c. make

>       d. Copy ``arch/x86/boot/bzImage`` to the ACRNSOS (L1 guest). Need this to launch the User VM (L2 guest)

Reviewer: I'm not 100% clear what this means, can you clarify?

(Note: I see later you use that file from `/home/guestl1/acrn-dm-bins/bzImage_uos` which is on the Service Vm rootfs. Is that what we should instruct the user to do here?

Author: Reworded, please check if this clarifies.

   .. note::
      Since the needed drivers are part of the User VM kernel itself, we can avoid installing the modules.

5. Build ubuntu.img using :ref:`build-the-ubuntu-kvm-image` and copy it to the ACRNSOS (L1 Guest).
   Alternatively you can also use virt-install to create a User VM image similar to ACRNSOS as shown below,

   .. code-block:: none

      virt-install \
      --name UOS \
      --ram 2048 \
      --disk path=/var/lib/libvirt/images/UOSUbuntu.img,size=8 \
      --vcpus 2 \
      --virt-type kvm \
      --os-type linux \
      --os-variant ubuntu18.04 \
      --graphics none \
      --location 'http://archive.ubuntu.com/ubuntu/dists/bionic/main/installer-amd64/' \
      --extra-args "console=tty0 console=ttyS0,115200n8"

   .. note::

>       Image at ``/var/lib/libvirt/images/UOSUbuntu.img`` is a qcom2 image. Convert it to raw image using, ``qemu-img convert -f qcow2 UOSUbuntu.img -O raw UOS.img``

Reviewer: ```suggestion
      Image at ``/var/lib/libvirt/images/UOSUbuntu.img`` is a qcow2 image. Convert it to raw image using, ``qemu-img convert -f qcow2 UOSUbuntu.img -O raw UOS.img``
```

> 6. Launch User VM using `launch_ubuntu.sh <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/devicemodel/samples/launch_ubuntu.sh>`_ script. Make sure to update with your ubuntu image and rootfs

Reviewer: Better point at the file in the repo that was cloned earlier.

This setup was tested with the following configuration,

- ACRN Hypervisor: tag acrn-2020w19.5-140000p (Commit ID: b0ae9cfa2b374d3faad9f890ff61e88f97205ac8)
- ACRN Kernel: 5.4.28 (Commit ID: a8cd22f49f0b2b56e526150fe0aaa9118edfcede)
- QEMU emulator version 4.2.0
- Service VM/User VM is ubuntu18.04
- Platforms Tested: ApolloLake, KabyLake, CoffeeLake

Pre-Requisites
**************
1. Make sure the platform supports Intel VMX as well as VT-d technologies. On ubuntu18.04, this
   can be checked by installing ``cpu-checker`` tool. If the output displays **KVM acceleration can be used**
   the platform supports it.

   .. code-block:: none

      $ kvm-ok
      INFO: /dev/kvm exists
      KVM acceleration can be used

2. Ensure the Ubuntu18.04 Host kernel version is **atleast 5.3.0** and above.

3. Make sure KVM and the following utilities are installed.

   .. code-block:: none

      $ sudo apt update && sudo apt upgrade -y
      $ sudo apt install qemu-kvm libvirt-bin virtinst -y

Prepare Service VM (L1 Guest)
*****************************
1. Use ``virt-install`` command to create Service VM.

   .. code-block:: none

>       virt-install \

Reviewer: ```suggestion
      $ virt-install \
```

Reviewer: Do you need `sudo` as well here?

         recordfail
         load_video
         gfxmode $linux_gfx_mode
         insmod gzio
         insmod part_msdos
         insmod ext2

         echo 'Loading ACRN hypervisor with SDC scenario ...'
         multiboot --quirk-modules-after-kernel /boot/acrn.32.out
         module /boot/bzImage_sos Linux_bzImage
      }

6. Update GRUB ``sudo update-grub``.

7. Shutdown the guest and relaunch using, ``virsh start ACRNSOS --console`` and select ACRN hypervisor from GRUB menu to launch Service VM running on top of ACRN.
   This can be verified from ``dmesg`` as shown below,

   .. code-block:: console

      guestl1@ACRNSOS:~$ dmesg | grep ACRN
      [    0.000000] Hypervisor detected: ACRN
      [    2.337176] ACRNTrace: Initialized acrn trace module with 4 cpu
      [    2.368358] ACRN HVLog: Initialized hvlog module with 4 cpu
      [    2.727905] systemd[1]: Set hostname to <ACRNSOS>.

8. When shutting down, make sure to cleanly destroy the Service VM to prevent crashes in subsequent boots. This can be done using,

   .. code-block:: none

>       virsh destroy ACRNSOS # where ACRNSOS is the virsh domain name.

Reviewer: ```suggestion
      $ virsh destroy ACRNSOS # where ACRNSOS is the virsh domain name.
```

Reviewer: Do we need `sudo` as well here?

Author: sudo is not required.

2. On the ACRNSOS guest, install shared libraries for acrn-dm (if not already installed).

   .. code-block:: none

      $ sudo apt-get install libpciaccess-dev

3. Install latest `IASL tool <https://acpica.org/downloads>`_ and copy the binary to ``/usr/sbin/iasl``.
   For this setup, used IASL 20200326 version but anything after 20190215 should be good.

4. Clone latest stable version or master and build ACRN User VM Kernel.

   .. code-block:: none

      $ git clone https://github.com/projectacrn/acrn-kernel
      $ cd acrn-kernel
      $ cp kernel_config_uos to .config
      $ make

   Once the User VM kernel is built successfully, copy ``arch/x86/boot/bzImage`` to  ACRNSOS (L1 guest) and rename this to ``bzImage_uos``. Need this to launch the User VM (L2 guest)

   .. note::
      The User VM kernel contains all needed drivers so you won't need to install extra kernel modules.

5. Build ubuntu.img using :ref:`build-the-ubuntu-kvm-image` and copy it to the ACRNSOS (L1 Guest).
   Alternatively you can also use virt-install to create a User VM image similar to ACRNSOS as shown below,

   .. code-block:: none

>       virt-install \

Reviewer: ```suggestion
      $ virt-install \
```

Reviewer: Do you need `sudo` to run `virt-install`?

Author: sudo is not required.

      --name UOS \
      --ram 2048 \
      --disk path=/var/lib/libvirt/images/UOSUbuntu.img,size=8 \
      --vcpus 2 \
      --virt-type kvm \
      --os-type linux \
      --os-variant ubuntu18.04 \
      --graphics none \
      --location 'http://archive.ubuntu.com/ubuntu/dists/bionic/main/installer-amd64/' \
      --extra-args "console=tty0 console=ttyS0,115200n8"

   .. note::
      Image at ``/var/lib/libvirt/images/UOSUbuntu.img`` is a qcow2 image. Convert it to raw image using, ``qemu-img convert -f qcow2 UOSUbuntu.img -O raw UOS.img``

6. Launch User VM using launch script from the cloned repo path ``devicemodel/samples/launch_ubuntu.sh``. Make sure to update with your ubuntu image and rootfs

   .. code-block:: none

>       acrn-dm -A -m $mem_size -s 0:0,hostbridge \

Reviewer: ```suggestion
      $ sudo acrn-dm -A -m $mem_size -s 0:0,hostbridge \
```

Author: This isn't a command for the user to run directly. This is just to inform the user to update the path to image as well as kernel. So ``sudo`` isn't required here.

Reviewer: Oh yes... you're right, please discard my comment!
